lsp-inline-completion.el

```diff
@@ -0,0 +1,365 @@
+  :type 'hook
+  :group 'lsp-mode)
+
+(defcustom lsp-inline-completion-shown-hook nil
+  "Hooks executed after showing a suggestion."
+  :type 'hook
+  :group 'lsp-mode)
+
+(defsubst lsp-inline-completion--overlay-visible ()
+  "Return whether the `overlay' is avaiable."
+  (and (overlayp lsp-inline-completion--overlay)
+       (overlay-buffer lsp-inline-completion--overlay)))
+
+(defun lsp-inline-completion--clear-overlay ()
+  "Hide the suggestion overlay"
+  (when (lsp-inline-completion--overlay-visible)
+    (delete-overlay lsp-inline-completion--overlay))
+  (setq lsp-inline-completion--overlay nil))
+
+
+(defun lsp-inline-completion--get-overlay (beg end)
+  "Build the suggestions overlay"
+  (when (overlayp lsp-inline-completion--overlay)
+    (lsp-inline-completion--clear-overlay))
+
+  (setq lsp-inline-completion--overlay (make-overlay beg end nil nil t))
+  (overlay-put lsp-inline-completion--overlay 'keymap lsp-inline-completion-active-map)
+  (overlay-put lsp-inline-completion--overlay 'priority 9000)
+
+  lsp-inline-completion--overlay)
+
+
+;;;###autoload
+(defun lsp-inline-completion-show-keys ()
+  "Shows active keymap hints in the minibuffer"
+
+  (unless (and lsp-inline-completion--items
+               (numberp lsp-inline-completion--current))
+    (error "No completions to show 1"))
+
+  (let ((message-log-max nil))
+    (message (concat "Completion "
+                     (propertize (format "%d" (1+ lsp-inline-completion--current)) 'face 'bold)
+                     "/"
+                     (propertize (format "%d" (length lsp-inline-completion--items)) 'face 'bold)
+
+                     (-when-let (keys (where-is-internal #'lsp-inline-completion-next lsp-inline-completion-active-map))
+                       (concat ". "
+                               (propertize " Next" 'face 'italic)
+                               (format ": [%s]"
+                                       (string-join (--map (propertize (key-description it) 'face 'help-key-binding)
+                                                           keys)
+                                                    "/"))))
+                     (-when-let (keys (where-is-internal #'lsp-inline-completion-accept lsp-inline-completion-active-map))
+                       (concat (propertize " Accept" 'face 'italic)
+                               (format ": [%s]"
+                                       (string-join (--map (propertize (key-description it) 'face 'help-key-binding)
+                                                           keys)
+                                                    "/"))))))))
+
+(defun lsp-inline-completion-show-overlay ()
+  "Makes the suggestion overlay visible"
+  (unless (and lsp-inline-completion--items
+               (numberp lsp-inline-completion--current))
+    (error "No completions to show 2"))
+
+  (lsp-inline-completion--clear-overlay)
+
+  (setq lsp-inline-completion--showing-company
```

Comment:
I wonder if we want to include `company-mode` integration here as we can have multiple alternative package provide the same functionality as `company-mode`, `corfu` for example.

What will happen if we don't cancel company mode here? The overlay from `company-mode` will be displayed over the inline completion? Can that be mitigate by making the overlay from inline completion topmost or bottommost?

Also, we can put it as a tip for the user (a simple function that the user can put to their init file) instead? Or even suggest the user to use a different company-mode's front end?
I've been using `copilot.el` with `company-mode` and haven't really notice issue so far though.

Reply:
> I wonder if we want to include company-mode integration here as we can have multiple alternative package provide the same functionality as company-mode, corfu for example.

Hmmm I agree that having no way of plugging other completion frontends may be an issue -- corfu users would have to advise the inline completion functions to have a similar level of integration, so not very ideal.

> What will happen if we don't cancel company mode here?

It depends on what frontends will be active at the time we display the overlay. With the default company tooltip frontend active, I only managed to get one-liners such as in the video below. It does not obsure the company completion list, but pressing tab or whatever uses the inline completion overlay active map, so this would cause some confusion on the user side.

[Kooha-2024-11-27-15-58-38.webm](https://github.com/user-attachments/assets/df712989-5f55-428b-97df-3139b329381a)

What happens if company-preview frontend is active? Not sure, probably the inline completion would obscure the preview completely.

If the user uses company-box or company-posframe (my case), then overlay priority makes no difference because company displays completions in a child frame :/

>  a tip for the user (a simple function that the user can put to their init file) instead? 

Good one!

I've refactored the company integration into a minor mode. The user can then set up their code such as in the following snippet:

```elisp
  (require 'lsp-inline-completion)
  ;; Show the current/total plus keys in the 
  (add-hook 'lsp-inline-completion-shown-hook #'lsp-inline-completion-show-keys)

  ;; Make inline completions play nice with company mode
  (add-hook 'lsp-inline-completion-mode-hook #'lsp-inline-completion-company-integration-mode)
```

Reply:
I like the Emacs binding. @kassick 
A question about `down-mouse ...` do we need to bind them to `#ignore`? Shouldn't that affect other minor-modes key binding?

Reply:
> I like the Emacs binding. @kassick

:+1: 

> A question about down-mouse ... do we need to bind them to #ignore? 

We do need them -- down-mouse-1 is sent before mouse-1. Without the ignore, this event is then captured by the `[t]` in the keymap , which cancels the completion and then checks what was bound in the currently active keymap and executes it.

So we do need to ignore down-mouse-1 if we want mouse-1 to accept the completion. I found this to be the case -- click somewhere in the overlay, have it accepted.

(I did not manage to get the cursor positioned in the correct place afterwards, mouse events in emacs are sourcery to me)

> Shouldn't that affect other minor-modes key binding?

It doesn't affect any other mode because the keymap is only active when the cursor is inside of the overlay -- it's not a minor mode keymap that would override any other mapping.

Reply:
I've updated the keymap accordingly.

Reply:
RE: _(I did not manage to get the cursor positioned in the correct place afterwards, mouse events in emacs are sourcery to me)_

I've been successfully nerd-snipped by myself ...

Clicking in the completion now accepts it and places the cursor in the correct position.